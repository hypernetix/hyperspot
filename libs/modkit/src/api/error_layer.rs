//! Centralized error mapping for Axum
//!
//! This module provides utilities for automatically converting all framework
//! and module errors into consistent RFC 9457 Problem+JSON responses, eliminating
//! per-route boilerplate.

use axum::{extract::Request, http::HeaderMap, middleware::Next, response::Response};
use std::any::Any;

use crate::api::problem::Problem;
use crate::context::ConfigError;
use modkit_odata::Error as ODataError;

/// Middleware function that provides centralized error mapping
///
/// This middleware can be applied to routes to automatically extract request context
/// and provide it to error handlers. The actual error conversion happens in the
/// `IntoProblem` trait implementations and `map_error_to_problem` function.
pub async fn error_mapping_middleware(request: Request, next: Next) -> Response {
    let _uri = request.uri().clone();
    let _headers = request.headers().clone();

    let response = next.run(request).await;

    // If the response is already successful or is already a Problem response, pass it through
    if response.status().is_success() || is_problem_response(&response) {
        return response;
    }

    // For error responses, the actual error conversion should happen in the handlers
    // using the IntoProblem trait or map_error_to_problem function
    // This middleware provides the infrastructure for extracting request context
    response
}

/// Check if a response is already a Problem+JSON response
fn is_problem_response(response: &Response) -> bool {
    response
        .headers()
        .get(axum::http::header::CONTENT_TYPE)
        .and_then(|v| v.to_str().ok())
        .map(|ct| ct.contains("application/problem+json"))
        .unwrap_or(false)
}

/// Extract trace ID from headers or generate one
pub fn extract_trace_id(headers: &HeaderMap) -> Option<String> {
    // Try to get trace ID from various common headers
    headers
        .get("x-trace-id")
        .or_else(|| headers.get("x-request-id"))
        .or_else(|| headers.get("traceparent"))
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string())
        .or_else(|| {
            // Try to get from current tracing span
            tracing::Span::current()
                .id()
                .map(|id| id.into_u64().to_string())
        })
}

/// Centralized error mapping function
///
/// This function provides a single place to convert all framework and module errors
/// into consistent Problem responses with proper trace IDs and instance paths.
pub fn map_error_to_problem(error: &dyn Any, instance: &str, trace_id: Option<String>) -> Problem {
    // Try to downcast to known error types
    if let Some(odata_err) = error.downcast_ref::<ODataError>() {
        return crate::api::odata::error::odata_error_to_problem(odata_err, instance, trace_id);
    }

    if let Some(config_err) = error.downcast_ref::<ConfigError>() {
        let mut problem = match config_err {
            ConfigError::ModuleNotFound { module } => Problem::new(
                500,
                "Configuration Error",
                format!("Module '{}' configuration not found", module),
            )
            .with_code("CONFIG_MODULE_NOT_FOUND")
            .with_type("https://errors.example.com/CONFIG_MODULE_NOT_FOUND"),

            ConfigError::InvalidModuleStructure { module } => Problem::new(
                500,
                "Configuration Error",
                format!("Module '{}' has invalid configuration structure", module),
            )
            .with_code("CONFIG_INVALID_STRUCTURE")
            .with_type("https://errors.example.com/CONFIG_INVALID_STRUCTURE"),

            ConfigError::MissingConfigSection { module } => Problem::new(
                500,
                "Configuration Error",
                format!("Module '{}' is missing required config section", module),
            )
            .with_code("CONFIG_MISSING_SECTION")
            .with_type("https://errors.example.com/CONFIG_MISSING_SECTION"),

            ConfigError::InvalidConfig { module, .. } => Problem::new(
                500,
                "Configuration Error",
                format!("Module '{}' has invalid configuration", module),
            )
            .with_code("CONFIG_INVALID")
            .with_type("https://errors.example.com/CONFIG_INVALID"),
        };

        problem = problem.with_instance(instance);
        if let Some(tid) = trace_id {
            problem = problem.with_trace_id(tid);
        }
        return problem;
    }

    // Handle anyhow::Error
    if let Some(anyhow_err) = error.downcast_ref::<anyhow::Error>() {
        let mut problem = Problem::new(500, "Internal Server Error", "An internal error occurred")
            .with_code("INTERNAL_ERROR")
            .with_type("https://errors.example.com/INTERNAL_ERROR");

        problem = problem.with_instance(instance);
        if let Some(tid) = trace_id {
            problem = problem.with_trace_id(tid);
        }

        // Log the full error for debugging
        tracing::error!(error = %anyhow_err, "Internal server error");
        return problem;
    }

    // Fallback for unknown error types
    let mut problem = Problem::new(500, "Unknown Error", "An unknown error occurred")
        .with_code("UNKNOWN_ERROR")
        .with_type("https://errors.example.com/UNKNOWN_ERROR");

    problem = problem.with_instance(instance);
    if let Some(tid) = trace_id {
        problem = problem.with_trace_id(tid);
    }

    tracing::error!("Unknown error type in error mapping layer");
    problem
}

/// Helper trait for converting errors to Problem responses with context
pub trait IntoProblem {
    fn into_problem(self, instance: &str, trace_id: Option<String>) -> Problem;
}

impl IntoProblem for ODataError {
    fn into_problem(self, instance: &str, trace_id: Option<String>) -> Problem {
        crate::api::odata::error::odata_error_to_problem(&self, instance, trace_id)
    }
}

impl IntoProblem for ConfigError {
    fn into_problem(self, instance: &str, trace_id: Option<String>) -> Problem {
        map_error_to_problem(&self as &dyn Any, instance, trace_id)
    }
}

impl IntoProblem for anyhow::Error {
    fn into_problem(self, instance: &str, trace_id: Option<String>) -> Problem {
        map_error_to_problem(&self as &dyn Any, instance, trace_id)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_odata_error_mapping() {
        let error = ODataError::InvalidFilter("malformed".to_string());
        let problem = error.into_problem("/test", Some("trace123".to_string()));

        assert_eq!(problem.status, 422);
        assert!(problem.code.contains("invalid_filter"));
        assert_eq!(problem.instance, "/test");
        assert_eq!(problem.trace_id, Some("trace123".to_string()));
    }

    #[test]
    fn test_config_error_mapping() {
        let error = ConfigError::ModuleNotFound {
            module: "test_module".to_string(),
        };
        let problem = error.into_problem("/api/test", None);

        assert_eq!(problem.status, 500);
        assert_eq!(problem.code, "CONFIG_MODULE_NOT_FOUND");
        assert_eq!(problem.instance, "/api/test");
        assert!(problem.detail.contains("test_module"));
    }

    #[test]
    fn test_anyhow_error_mapping() {
        let error = anyhow::anyhow!("Something went wrong");
        let problem = error.into_problem("/api/test", Some("trace456".to_string()));

        assert_eq!(problem.status, 500);
        assert_eq!(problem.code, "INTERNAL_ERROR");
        assert_eq!(problem.instance, "/api/test");
        assert_eq!(problem.trace_id, Some("trace456".to_string()));
    }

    #[test]
    fn test_extract_trace_id_from_headers() {
        let mut headers = HeaderMap::new();
        headers.insert("x-trace-id", "test-trace-123".parse().unwrap());

        let trace_id = extract_trace_id(&headers);
        assert_eq!(trace_id, Some("test-trace-123".to_string()));
    }
}
