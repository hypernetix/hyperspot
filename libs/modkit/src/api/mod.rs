//! Type-safe API operation builder with compile-time guarantees
//!
//! This module provides a type-state builder pattern that enforces at compile time
//! that API operations cannot be registered unless both a handler and at least one
//! response are specified.

pub mod operation_builder;

pub use operation_builder::{
    state, Missing, OpenApiRegistry, OperationBuilder, OperationSpec, ParamLocation, ParamSpec,
    Present, ResponseSpec, ensure_schema,
};
