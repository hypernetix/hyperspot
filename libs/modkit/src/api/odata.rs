use axum::extract::{FromRequestParts, Query};
use axum::http::request::Parts;
use modkit_odata::{ast, CursorV1, Error as ODataError, ODataOrderBy, OrderKey, SortDir};
use odata_params::filters as od;
use serde::Deserialize;

// Re-export types from modkit-odata for convenience and better DX
pub use modkit_odata::ODataQuery;
// CursorV1 is available through the private import above for internal use

// Re-export error mapping from the error module
pub mod error;
pub use error::odata_error_to_problem;

#[derive(Deserialize, Default)]
pub struct ODataParams {
    #[serde(rename = "$filter")]
    pub filter: Option<String>,
    #[serde(rename = "$orderby")]
    pub orderby: Option<String>,
    pub limit: Option<u64>,
    pub cursor: Option<String>,
}

pub const MAX_FILTER_LEN: usize = 8 * 1024;
pub const MAX_NODES: usize = 2000;
pub const MAX_ORDERBY_LEN: usize = 1024;
pub const MAX_ORDER_FIELDS: usize = 10;

/// Parse $orderby string into ODataOrderBy
/// Format: "field1 [asc|desc], field2 [asc|desc], ..."
/// Default direction is asc if not specified
pub fn parse_orderby(raw: &str) -> Result<ODataOrderBy, modkit_odata::Error> {
    let raw = raw.trim();
    if raw.is_empty() {
        return Ok(ODataOrderBy::empty());
    }

    if raw.len() > MAX_ORDERBY_LEN {
        return Err(modkit_odata::Error::InvalidOrderByField(
            "orderby too long".into(),
        ));
    }

    let mut keys = Vec::new();

    for part in raw.split(',') {
        let part = part.trim();
        if part.is_empty() {
            continue;
        }

        let tokens: Vec<&str> = part.split_whitespace().collect();
        let (field, dir) = match tokens.as_slice() {
            [field] => (*field, SortDir::Asc),
            [field, "asc"] => (*field, SortDir::Asc),
            [field, "desc"] => (*field, SortDir::Desc),
            _ => {
                return Err(modkit_odata::Error::InvalidOrderByField(format!(
                    "invalid orderby clause: {}",
                    part
                )))
            }
        };

        if field.is_empty() {
            return Err(modkit_odata::Error::InvalidOrderByField(
                "empty field name in orderby".into(),
            ));
        }

        keys.push(OrderKey {
            field: field.to_string(),
            dir,
        });
    }

    if keys.len() > MAX_ORDER_FIELDS {
        return Err(modkit_odata::Error::InvalidOrderByField(
            "too many order fields".into(),
        ));
    }

    Ok(ODataOrderBy(keys))
}

/// Extract and validate full OData query from request parts
/// - Parses $filter, $orderby, limit, cursor
/// - Enforces budgets and validates formats
/// - Returns unified ODataQuery
pub async fn extract_odata_query<S>(
    parts: &mut Parts,
    state: &S,
) -> Result<ODataQuery, crate::api::problem::Problem>
where
    S: Send + Sync,
{
    // Parse query; default if missing
    let Query(params) = Query::<ODataParams>::from_request_parts(parts, state)
        .await
        .unwrap_or_else(|_| Query(ODataParams::default()));

    let mut query = ODataQuery::new();

    // Parse filter
    if let Some(raw_filter) = params.filter.as_ref() {
        let raw = raw_filter.trim();
        if !raw.is_empty() {
            if raw.len() > MAX_FILTER_LEN {
                return Err(crate::api::bad_request("Filter too long"));
            }

            // Parse into odata_params AST
            let ast_src = od::parse_str(raw)
                .map_err(|e| crate::api::bad_request(format!("invalid $filter: {:?}", e)))?;

            // Complexity budget (node count)
            fn count_nodes(e: &od::Expr) -> usize {
                use od::Expr::*;
                match e {
                    Value(_) | Identifier(_) => 1,
                    Not(x) => 1 + count_nodes(x),
                    And(a, b) | Or(a, b) | Compare(a, _, b) => 1 + count_nodes(a) + count_nodes(b),
                    In(a, list) => 1 + count_nodes(a) + list.iter().map(count_nodes).sum::<usize>(),
                    Function(_, args) => 1 + args.iter().map(count_nodes).sum::<usize>(),
                }
            }
            if count_nodes(&ast_src) > MAX_NODES {
                return Err(crate::api::bad_request("Filter too complex"));
            }

            // Convert to transport-agnostic core AST
            let core_expr: ast::Expr = ast_src.into();

            // Generate filter hash for cursor consistency
            let filter_hash = modkit_odata::pagination::short_filter_hash(Some(&core_expr));

            query = query.with_filter(core_expr);
            if let Some(hash) = filter_hash {
                query = query.with_filter_hash(hash);
            }
        }
    }

    // Check for cursor+orderby conflict before parsing either
    if params.cursor.is_some() && params.orderby.is_some() {
        return Err(crate::api::odata::odata_error_to_problem(
            &ODataError::OrderWithCursor,
            "/",
            None,
        ));
    }

    // Parse cursor first (if present, skip orderby)
    if let Some(cursor_str) = params.cursor.as_ref() {
        let cursor = CursorV1::decode(cursor_str).map_err(|_| {
            crate::api::odata::odata_error_to_problem(&ODataError::InvalidCursor, "/", None)
        })?;
        query = query.with_cursor(cursor);
        // When cursor is present, order is empty (derived from cursor.s later)
        query = query.with_order(ODataOrderBy::empty());
    } else {
        // Parse orderby only when cursor is absent
        if let Some(raw_orderby) = params.orderby.as_ref() {
            let order = parse_orderby(raw_orderby)
                .map_err(|e| crate::api::odata::odata_error_to_problem(&e, "/", None))?;
            query = query.with_order(order);
        }
    }

    // Parse limit
    if let Some(limit) = params.limit {
        if limit == 0 {
            return Err(crate::api::odata::odata_error_to_problem(
                &ODataError::InvalidLimit,
                "/",
                None,
            ));
        }
        query = query.with_limit(limit);
    }

    Ok(query)
}

use std::ops::Deref;

/// Simple Axum extractor for full OData query parameters.
/// Parses $filter, $orderby, limit, and cursor parameters.
/// Usage in handlers:
///   async fn list_users(OData(query): OData, /* ... */) { /* use `query` */ }
#[derive(Debug, Clone)]
pub struct OData(pub ODataQuery);

impl OData {
    #[inline]
    pub fn into_inner(self) -> ODataQuery {
        self.0
    }
}

impl Deref for OData {
    type Target = ODataQuery;
    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl AsRef<ODataQuery> for OData {
    #[inline]
    fn as_ref(&self) -> &ODataQuery {
        &self.0
    }
}

impl From<OData> for ODataQuery {
    #[inline]
    fn from(x: OData) -> Self {
        x.0
    }
}

impl<S> FromRequestParts<S> for OData
where
    S: Send + Sync,
{
    type Rejection = crate::api::problem::Problem;

    #[allow(clippy::manual_async_fn)]
    fn from_request_parts(
        parts: &mut Parts,
        state: &S,
    ) -> impl core::future::Future<Output = Result<Self, Self::Rejection>> + Send {
        async move {
            let query = extract_odata_query(parts, state).await?;
            Ok(OData(query))
        }
    }
}

#[cfg(test)]
#[path = "odata_tests.rs"]
mod odata_tests;
