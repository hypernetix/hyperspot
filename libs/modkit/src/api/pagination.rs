//! Filter hashing utilities for OData pagination

use hex;
use odata_core::ast;
use sha2::{Digest, Sha256};

/// Normalize filter AST for consistent hashing
/// Produces a stable string representation for deterministic hashing
#[must_use]
pub fn normalize_filter_for_hash(expr: &ast::Expr) -> String {
    fn normalize_expr(expr: &ast::Expr) -> String {
        match expr {
            ast::Expr::And(left, right) => {
                format!("AND({},{})", normalize_expr(left), normalize_expr(right))
            }
            ast::Expr::Or(left, right) => {
                format!("OR({},{})", normalize_expr(left), normalize_expr(right))
            }
            ast::Expr::Not(inner) => {
                format!("NOT({})", normalize_expr(inner))
            }
            ast::Expr::Compare(left, op, right) => {
                let op_str = match op {
                    ast::CompareOperator::Eq => "EQ",
                    ast::CompareOperator::Ne => "NE",
                    ast::CompareOperator::Gt => "GT",
                    ast::CompareOperator::Ge => "GE",
                    ast::CompareOperator::Lt => "LT",
                    ast::CompareOperator::Le => "LE",
                };
                format!(
                    "CMP({},{},{})",
                    normalize_expr(left),
                    op_str,
                    normalize_expr(right)
                )
            }
            ast::Expr::In(expr, list) => {
                let list_str = list
                    .iter()
                    .map(normalize_expr)
                    .collect::<Vec<_>>()
                    .join(",");
                format!("IN({},{})", normalize_expr(expr), list_str)
            }
            ast::Expr::Function(name, args) => {
                let args_str = args
                    .iter()
                    .map(normalize_expr)
                    .collect::<Vec<_>>()
                    .join(",");
                format!("FN({},{})", name.to_lowercase(), args_str)
            }
            ast::Expr::Identifier(name) => {
                format!("ID({})", name.to_lowercase())
            }
            ast::Expr::Value(value) => match value {
                ast::Value::Null => "NULL".to_string(),
                ast::Value::Bool(b) => format!("BOOL({})", b),
                ast::Value::Number(n) => format!("NUM({})", n.normalized()),
                ast::Value::Uuid(u) => {
                    format!("UUID({})", u.as_hyphenated().to_string().to_lowercase())
                }
                ast::Value::DateTime(dt) => format!("DATETIME({})", dt.to_rfc3339()),
                ast::Value::Date(d) => format!("DATE({})", d.format("%Y-%m-%d")),
                ast::Value::Time(t) => format!("TIME({})", t.format("%H:%M:%S%.f")),
                ast::Value::String(s) => format!("STR({})", s),
            },
        }
    }

    normalize_expr(expr)
}

/// Generate a short hash from a filter expression for cursor consistency checks
/// Returns a 16-character hex string (64-bit hash)
#[must_use]
pub fn short_filter_hash(expr: Option<&ast::Expr>) -> Option<String> {
    expr.map(|e| {
        let normalized = normalize_filter_for_hash(e);
        let mut hasher = Sha256::new();
        hasher.update(normalized.as_bytes());
        let bytes = hasher.finalize();
        hex::encode(&bytes[..8]) // Take first 8 bytes for 64-bit hash
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use odata_core::ast::{CompareOperator, Expr, Value};

    #[test]
    fn test_normalize_filter_consistency() {
        // Test that the same logical filter produces the same normalized string
        let expr1 = Expr::Compare(
            Box::new(Expr::Identifier("name".to_string())),
            CompareOperator::Eq,
            Box::new(Expr::Value(Value::String("test".to_string()))),
        );

        let expr2 = Expr::Compare(
            Box::new(Expr::Identifier("name".to_string())),
            CompareOperator::Eq,
            Box::new(Expr::Value(Value::String("test".to_string()))),
        );

        assert_eq!(
            normalize_filter_for_hash(&expr1),
            normalize_filter_for_hash(&expr2)
        );
    }

    #[test]
    fn test_short_filter_hash_consistency() {
        let expr = Expr::Compare(
            Box::new(Expr::Identifier("id".to_string())),
            CompareOperator::Gt,
            Box::new(Expr::Value(Value::Number(42.into()))),
        );

        let hash1 = short_filter_hash(Some(&expr));
        let hash2 = short_filter_hash(Some(&expr));

        assert_eq!(hash1, hash2);
        assert!(hash1.is_some());
        assert_eq!(hash1.as_ref().unwrap().len(), 16); // 8 bytes = 16 hex chars
    }

    #[test]
    fn test_short_filter_hash_none() {
        assert_eq!(short_filter_hash(None), None);
    }
}
