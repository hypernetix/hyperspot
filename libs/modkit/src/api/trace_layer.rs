//! Trace propagation utilities for Problem responses
//!
//! This module provides helper traits and functions to automatically enrich
//! `Problem` with trace context:
//! - `trace_id`: extracted from the current tracing span
//! - `instance`: extracted from the request URI
//!
//! This eliminates per-callsite boilerplate and ensures consistent error reporting.

use crate::api::problem::Problem;

/// Extract trace_id from the current tracing span
fn extract_trace_id() -> Option<String> {
    // Try to extract from the current span's trace_id field
    // This requires coordination with the tracing subscriber
    tracing::Span::current().id().map(|id| format!("{:?}", id))
}

/// Helper trait for enriching Problem with trace context
pub trait WithTraceContext {
    /// Enrich this Problem with trace_id and instance from the current request context
    fn with_trace_context(self, instance: impl Into<String>) -> Self;
}

impl WithTraceContext for Problem {
    fn with_trace_context(mut self, instance: impl Into<String>) -> Self {
        self = self.with_instance(instance);
        if let Some(tid) = extract_trace_id() {
            self = self.with_trace_id(tid);
        }
        self
    }
}

/// Middleware-friendly: enrich errors from Axum extractors
///
/// Use this in handlers to automatically add trace context:
///
/// ```ignore
/// async fn handler(uri: Uri) -> Result<Json<Data>, Problem> {
///     let data = fetch_data()
///         .await
///         .map_err(Problem::from)
///         .map_err(|p| p.with_request_context(&uri))?;
///     Ok(Json(data))
/// }
/// ```
pub trait WithRequestContext {
    /// Add trace_id and instance from the current request
    fn with_request_context(self, uri: &axum::http::Uri) -> Self;
}

impl WithRequestContext for Problem {
    fn with_request_context(self, uri: &axum::http::Uri) -> Self {
        self.with_trace_context(uri.path())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_with_trace_context() {
        let problem = Problem::new(404, "Not Found", "Resource not found")
            .with_trace_context("/api/users/123");

        assert_eq!(problem.instance, "/api/users/123");
        // trace_id may or may not be set depending on tracing context
    }

    #[test]
    fn test_with_request_context() {
        use axum::http::Uri;

        let uri: Uri = "/api/users/123".parse().unwrap();
        let problem =
            Problem::new(404, "Not Found", "Resource not found").with_request_context(&uri);

        assert_eq!(problem.instance, "/api/users/123");
    }
}
