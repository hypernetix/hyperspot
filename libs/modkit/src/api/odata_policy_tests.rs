//! Tests for cursor+orderby policy enforcement

#[cfg(test)]
mod tests {
    use super::super::odata::*;
    use axum::http::{request::Parts, Uri};
    use modkit_odata::{CursorV1, SortDir};

    fn mock_parts(query_string: &str) -> Parts {
        let uri: Uri = format!("http://example.com/test?{}", query_string)
            .parse()
            .unwrap();
        let request = axum::http::Request::builder().uri(uri).body(()).unwrap();
        let (parts, _) = request.into_parts();
        parts
    }

    #[tokio::test]
    async fn test_cursor_with_orderby_conflict() {
        let mut parts = mock_parts("cursor=dGVzdA%3D%3D&%24orderby=id%20desc");
        let result = extract_odata_query(&mut parts, &()).await;

        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }

    #[tokio::test]
    async fn test_cursor_only_success() {
        // Create a valid cursor
        let cursor = CursorV1 {
            k: vec!["test".to_string()],
            o: SortDir::Desc,
            s: "-id".to_string(),
            f: None,
        };
        let cursor_encoded = cursor.encode();

        let mut parts = mock_parts(&format!("cursor={}", urlencoding::encode(&cursor_encoded)));
        let result = extract_odata_query(&mut parts, &()).await;

        assert!(result.is_ok());
        let query = result.unwrap();
        assert!(query.cursor.is_some());
        assert!(query.order.is_empty()); // Order should be empty when cursor is present
    }

    #[tokio::test]
    async fn test_orderby_only_success() {
        let mut parts = mock_parts("%24orderby=id%20desc%2C%20name%20asc");
        let result = extract_odata_query(&mut parts, &()).await;

        assert!(result.is_ok());
        let query = result.unwrap();
        assert!(query.cursor.is_none());
        assert!(!query.order.is_empty());
        assert_eq!(query.order.0.len(), 2);
        assert_eq!(query.order.0[0].field, "id");
        assert_eq!(query.order.0[0].dir, SortDir::Desc);
        assert_eq!(query.order.0[1].field, "name");
        assert_eq!(query.order.0[1].dir, SortDir::Asc);
    }

    #[tokio::test]
    async fn test_neither_cursor_nor_orderby() {
        let mut parts = mock_parts("limit=10");
        let result = extract_odata_query(&mut parts, &()).await;

        assert!(result.is_ok());
        let query = result.unwrap();
        assert!(query.cursor.is_none());
        assert!(query.order.is_empty());
        assert_eq!(query.limit, Some(10));
    }

    #[tokio::test]
    async fn test_invalid_cursor_error() {
        let mut parts = mock_parts("cursor=invalid_base64");
        let result = extract_odata_query(&mut parts, &()).await;

        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }
}
