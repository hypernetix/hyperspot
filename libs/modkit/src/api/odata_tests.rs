#[cfg(test)]
mod tests {
    use crate::api::odata::*;
    use axum::extract::FromRequestParts;
    use axum::http::Request;

    #[test]
    fn test_parse_orderby_simple() {
        let result = parse_orderby("created_at desc").unwrap();
        assert_eq!(result.0.len(), 1);
        assert_eq!(result.0[0].field, "created_at");
        assert_eq!(result.0[0].dir, SortDir::Desc);
    }

    #[test]
    fn test_parse_orderby_multiple_fields() {
        let result = parse_orderby("created_at desc, id asc, name").unwrap();
        assert_eq!(result.0.len(), 3);

        assert_eq!(result.0[0].field, "created_at");
        assert_eq!(result.0[0].dir, SortDir::Desc);

        assert_eq!(result.0[1].field, "id");
        assert_eq!(result.0[1].dir, SortDir::Asc);

        assert_eq!(result.0[2].field, "name");
        assert_eq!(result.0[2].dir, SortDir::Asc); // default
    }

    #[test]
    fn test_parse_orderby_whitespace_tolerance() {
        let result = parse_orderby("  created_at   desc  ,   id   asc  ").unwrap();
        assert_eq!(result.0.len(), 2);
        assert_eq!(result.0[0].field, "created_at");
        assert_eq!(result.0[0].dir, SortDir::Desc);
        assert_eq!(result.0[1].field, "id");
        assert_eq!(result.0[1].dir, SortDir::Asc);
    }

    #[test]
    fn test_parse_orderby_empty() {
        let result = parse_orderby("").unwrap();
        assert!(result.is_empty());

        let result = parse_orderby("   ").unwrap();
        assert!(result.is_empty());
    }

    #[test]
    fn test_parse_orderby_too_long() {
        let long_orderby = "a".repeat(MAX_ORDERBY_LEN + 1);
        let result = parse_orderby(&long_orderby);
        assert!(result.is_err());
        assert!(matches!(
            result.unwrap_err(),
            modkit_odata::Error::InvalidOrderByField(_)
        ));
    }

    #[test]
    fn test_parse_orderby_too_many_fields() {
        let many_fields: Vec<String> = (0..=MAX_ORDER_FIELDS)
            .map(|i| format!("field{}", i))
            .collect();
        let orderby = many_fields.join(", ");

        let result = parse_orderby(&orderby);
        assert!(result.is_err());
        assert!(matches!(
            result.unwrap_err(),
            modkit_odata::Error::InvalidOrderByField(_)
        ));
    }

    #[test]
    fn test_parse_orderby_invalid_clause() {
        let result = parse_orderby("field invalid_direction");
        assert!(result.is_err());
        assert!(matches!(
            result.unwrap_err(),
            modkit_odata::Error::InvalidOrderByField(_)
        ));
    }

    #[test]
    fn test_parse_orderby_empty_field() {
        let result = parse_orderby(", asc");
        // The new implementation is more lenient and skips empty segments
        // so this now succeeds with one field "asc"
        assert!(result.is_ok());
        let order = result.unwrap();
        assert_eq!(order.0.len(), 1);
        assert_eq!(order.0[0].field, "asc");
    }

    #[tokio::test]
    async fn test_extract_odata_query_full() {
        let uri = "/?%24filter=email%20eq%20%27test%40example.com%27&%24orderby=created_at%20desc&limit=25&cursor=eyJ2IjoxLCJrIjpbInRlc3QiXSwicyI6Ii1jcmVhdGVkX2F0Iiwib28oImFzYyJ9";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let result = extract_odata_query(&mut parts, &()).await;
        assert!(result.is_err());
        let problem = result.unwrap_err();
        // Extract status from the problem for testing
        // OData errors return 422 (Unprocessable Entity) per GTS catalog
        assert_eq!(problem.status, 422);

        // Should have cursor (even if decode fails, it should try)
        // Note: The cursor in the test is not a valid base64url, but that's OK for this test
    }

    #[tokio::test]
    async fn test_extract_odata_query_filter_only() {
        let uri = "/?%24filter=email%20eq%20%27test%40example.com%27";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let query = extract_odata_query(&mut parts, &()).await.unwrap();

        assert!(query.filter.is_some());
        assert!(query.order.is_empty());
        assert_eq!(query.limit, None);
        assert!(query.cursor.is_none());
    }

    #[tokio::test]
    async fn test_extract_odata_query_orderby_only() {
        let uri = "/?%24orderby=created_at%20desc%2C%20id%20asc";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let query = extract_odata_query(&mut parts, &()).await.unwrap();

        assert!(query.filter.is_none());
        assert_eq!(query.order.0.len(), 2);
        assert_eq!(query.limit, None);
        assert!(query.cursor.is_none());
    }

    #[tokio::test]
    async fn test_extract_odata_query_empty() {
        let uri = "/";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let query = extract_odata_query(&mut parts, &()).await.unwrap();

        assert!(query.filter.is_none());
        assert!(query.order.is_empty());
        assert_eq!(query.limit, None);
        assert!(query.cursor.is_none());
    }

    #[tokio::test]
    async fn test_extract_odata_query_limit_zero_error() {
        let uri = "/?limit=0";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let result = extract_odata_query(&mut parts, &()).await;
        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }

    #[tokio::test]
    async fn test_extract_odata_query_filter_too_long() {
        let long_filter = "email eq '".to_string() + &"a".repeat(MAX_FILTER_LEN) + "'";
        let uri = format!("/?%24filter={}", urlencoding::encode(&long_filter));

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let result = extract_odata_query(&mut parts, &()).await;
        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }

    #[tokio::test]
    async fn test_extract_odata_query_invalid_filter() {
        let uri = "/?%24filter=invalid%20syntax%20here";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let result = extract_odata_query(&mut parts, &()).await;
        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }

    #[tokio::test]
    async fn test_extract_odata_query_invalid_orderby() {
        let uri = "/?%24orderby=field%20invalid_direction";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let result = extract_odata_query(&mut parts, &()).await;
        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }

    #[tokio::test]
    async fn test_extract_odata_query_invalid_cursor() {
        let uri = "/?cursor=invalid_cursor";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let result = extract_odata_query(&mut parts, &()).await;
        assert!(result.is_err());
        let _problem_response = result.unwrap_err();
    }

    #[tokio::test]
    async fn test_odata_extractor() {
        let uri = "/?%24filter=email%20eq%20%27test%40example.com%27&limit=10";

        let request = Request::builder().uri(uri).body(()).unwrap();

        let (mut parts, _body) = request.into_parts();

        let odata = OData::from_request_parts(&mut parts, &()).await.unwrap();

        assert!(odata.filter.is_some());
        assert_eq!(odata.limit, Some(10));
    }

    #[test]
    fn test_odata_deref() {
        use modkit_odata::ast::*;

        let expr = Expr::Identifier("test".to_string());
        let query = ODataQuery::default().with_filter(expr);
        let odata = OData(query);

        // Test Deref
        assert!(odata.has_filter());

        // Test AsRef
        let query_ref: &ODataQuery = odata.as_ref();
        assert!(query_ref.has_filter());

        // Test Into
        let query_back: ODataQuery = odata.into();
        assert!(query_back.has_filter());
    }
}
