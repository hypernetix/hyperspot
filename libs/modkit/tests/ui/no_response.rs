//! This test should fail to compile because register() is called without a response

use modkit::api::OperationBuilder;
use axum::Router;

async fn test_handler() -> &'static str { "ok" }

struct DummyRegistry;
impl modkit::api::OpenApiRegistry for DummyRegistry {
    fn register_operation(&self, _: &modkit::api::OperationSpec) {}
    fn register_schema(&self, _: &str, _: schemars::schema::RootSchema) {}
    fn as_any(&self) -> &dyn std::any::Any { self }
}

fn main() {
    let registry = DummyRegistry;
    let router = Router::new();

    // This should fail to compile - missing response
    let _ = OperationBuilder::<_, _, ()>::get("/test")
        .summary("Test endpoint")
        .handler(test_handler)
        .register(router, &registry);
}
