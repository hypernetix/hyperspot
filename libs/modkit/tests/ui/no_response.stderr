error[E0599]: no method named `register` found for struct `modkit::OperationBuilder<Present, Missing, ()>` in the current scope
  --> tests/ui/no_response.rs:25:10
   |
22 |       let _ = OperationBuilder::<_, _, ()>::get("/test")
   |  _____________-
23 | |         .summary("Test endpoint")
24 | |         .handler(test_handler)
25 | |         .register(router, &registry);
   | |         -^^^^^^^^ method not found in `modkit::OperationBuilder<Present, Missing, ()>`
   | |_________|
   |
   |
   = note: the method was found for
           - `modkit::OperationBuilder<Present, Present, S>`
